### ISSUE TITLE
Story section URL problems

### ISSUE THREAD

**Issue by ElenaGallo** [geosolutions-it/MapStore2#5757]
### Issue n°1
The shared URL of a story section displays the URL of the **first** story section

**How to Reproduce**
- Open this story: https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/27159/7b688800-d58b-11ea-baa0-d1aae6c8107f
- Scroll down the story to the Banner section
- Click on Share 
- Enable the scrolling position 
- Copy the story link and open it in a new browser window

**Expected Result** 
The story should scroll to the Banner section at the URL https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/shared/27159/3d6ccdf1-2a63-446b-9372-297406a36d0d 

**Current Result**
The URL reported in the Share tool is correct: https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/shared/27159/3d6ccdf1-2a63-446b-9372-297406a36d0d 

But the link of the story becomes https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/shared/27159/7b688800-d58b-11ea-baa0-d1aae6c8107f without any scrolling to the Banner section

### Issue n°2
In the shared story, the UUIDs of each section are incorrect after clicking on NavBar to view a specific section

![share section1](https://user-images.githubusercontent.com/56537133/89559525-1451b280-d816-11ea-871d-7f8a0e1795f5.gif)

**How to Reproduce**
- On the Share story 
- Click for example on **Banner Section** on NavBar (same issue of issue n°6 below, the story does't scroll)
- Scroll the story

**Expected Result**  
The story URL changes based on the section you're viewing with a correct ID

**Current Result**
The story URL is not properly generated some time and the status of the navbar is not correct (items are not properly highlighted)

![2020_08_07_12_19_16_](https://user-images.githubusercontent.com/1280027/89636144-6e02bd00-d8a8-11ea-9d30-a8e7f36751ec.png)

### Issue n°3
Sometime the item in the navbar is not properly highlighted and a wrong section ID is reported in the URL **title_section_id1**

![2020_08_07_10_31_02_](https://user-images.githubusercontent.com/1280027/89626464-c8e0e800-d899-11ea-859c-42a23e04dd11.png)

**How to reproduce**

- Login as admin
- Open this story https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/11555/7b688800-d58b-11ea-baa0-d1aae6c8107f
- Go to homepage and search and select `test_sharesection` story

### Issue n°4

A wrong URL is generated for web page section:

**How to reproduce**
- Open this story: https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/27159/7b688800-d58b-11ea-baa0-d1aae6c8107f
- Click on "Media Section" in the story navbar (the Media Section item in the navbar must be properly selected)
- Click on "Web Page Section" in the navbar 

**Expected result**
The URL should be: https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/27159/12ee40ac-5644-49f2-bf18-7398691af53c
This is the URL generated if you click on "Web Page Section" without clicking on "Media Section" before

**Current result**
The generated URL is: https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/27159/12ee40ac-5644-49f2-bf18-7398691af53c/357bf2f3-00f0-4201-b796-fbb46283fcb4

Unfortunatelly it is not possible to always replicate this **randomic** issue that seems to be connected to a bug in the section ID definition logic for the URL 

### Issue n°5

A more frequent issue below:

**How to reproduce**
- Open this story: https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/27159/7b688800-d58b-11ea-baa0-d1aae6c8107f
- Click on "Media Section" in the story navbar 

**Expected result**
The story scroll to the Media Section properly and the item in the navbar is highlighted

**Current result**
The story scroll in a wrong position and the Web Page Section item is highlighted in the navbar

![story_url_5](https://user-images.githubusercontent.com/1280027/89794997-1ecabf80-db28-11ea-9256-1c502c1c573e.gif)

### Issue n°6

The navbar in the shared story URL is not working

**How to reproduce**
- Open this story: https://dev.mapstore2.geo-solutions.it/mapstore/#/geostory/shared/27159/7b688800-d58b-11ea-baa0-d1aae6c8107f
- Click on items in the navbar

**Expected result**
The story properly scroll to the desired position and the related item in the navbar is highlighted

**Current result**
The URL changes when you clink on an item in the navbar but the story doesn't scroll

![story_url_4](https://user-images.githubusercontent.com/1280027/89634675-0cd9ea00-d8a6-11ea-8851-6732284f22ac.gif)

## Other useful information

Related to #5569

**Comment by tdipisa** [comment]
Testing again the current implementation today I found also the following:

1. Sometime the item in the navbar (maybe only for the new Banner section?) is not properly highlighted and a wrong section ID is reported in the URL **title_section_id1**
![2020_08_07_10_31_02_](https://user-images.githubusercontent.com/1280027/89626464-c8e0e800-d899-11ea-859c-42a23e04dd11.png)

2. The video below shows also how currently the banner section looks not properly highlighed and a wrong empty space appears in the nav toolbar

**Comment by jstrR** [comment]
@tdipisa @allyoucanmap due to my investigation, issues 1,2,6 related to the problem with consecutive dynamic routes. That's why yo fix them, we need to change current url pattern from :gid/:sectionId/:columnId to the /:gid/section/:sectionId/column/:columnId or /:gid&sectionId&columnId

**Comment by tdipisa** [comment]
@jstrR I assume that you have already tested both cases and they are working properly. Therefore, I kindly ask you to use the following:

/:gid/section/:sectionId/column/:columnId

What about other issues 3, 4, 5? Are those issues also connected to this or they are independent?

**Comment by jstrR** [comment]
@tdipisa @allyoucanmap
n4
this story has column with related id and sometimes it gets on focus, thats why 357bf2f3-00f0-4201-b796-fbb46283fcb4 is added at the end
n5 
you click mediaSection -> page scrolls to the video cover. but right now you have loader instead of videocover, which smaller than initial cover. your video is loaded and because it has a lot of height it is lowered on the page. but currently your url has already been updated to the new specialized section.

**Comment by allyoucanmap** [comment]
@tdipisa @jstrR  about issue n 5

> you click mediaSection -> page scrolls to the video cover. but right now you have loader instead of videocover, which smaller than initial cover. your video is loaded and because it has a lot of height it is lowered on the page. but currently your url has already been updated to the new specialized section.

As commented by @jstrR this happens only because the size of the target media is changing after loading.

- so the scrollbar reaches the target correctly

![image](https://user-images.githubusercontent.com/19175505/89878173-79fcc080-dbc1-11ea-8387-efa8a89ab7f7.png)

- after a debounce time of 300ms the loading spinner is replaced by the actual video but the video has a different height

![image](https://user-images.githubusercontent.com/19175505/89878396-c6e09700-dbc1-11ea-903c-3fae4eefc937.png)

I tried to check if this behaviour could be fixed adjusting the debounce time of loader spinner but unfortunately it does not work and the result is worst than now.

We still need the debounce time to prevent other contents to load while scrolling the whole story.

I think this issue is not new and it's caused mainly by media contents inside column and paragraph where we don't know in advance the aspect ratio of content.

A solutions could be the use of loading placeholders with the same aspect ratio and size of the final content to ensure the structure of the story will be the same before and after loading. (this means we need to store this information in the story config someway)

Probably this issue needs to be investigated separately.

**Comment by tdipisa** [comment]
@jstrR I will create a separate issue for point n°5, thank you so much @allyoucanmap.

**Comment by allyoucanmap** [comment]
@jstrR @tdipisa  about issue n 4

This issue is reproducible only when you load the story the first time because the contents at the bottom are not loaded yet so their height will be different after the loading is complete (see [comment](https://github.com/geosolutions-it/MapStore2/issues/5757#issuecomment-671833274) for issue 5).

Currently the navigation bar highlights a nav item based on the content visible in the view in this way is possible to highlight it automatically while scrolling or performing other actions (eg: zoom to in the editing list).

This is the navbar workflow:

- click an nav item
- use the id of nav item to query in the DOM the correspondant section/column with same id
- if the node is available it uses the function [scrollIntoView](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView)
- eventually the document scroll the section/column into the view

Note: the clicked item of navbar is not stored in the state

This is the situation after clicking the last item in the navbar `Media Section`:

- the video is still loading so the height of loader is less than the actual video
- we have **three** sections visible in the window view: Web Page Section, Banner Section and Media Section

![image](https://user-images.githubusercontent.com/19175505/89996094-a6c8da80-dc8a-11ea-9a05-e01fc890cc1b.png)

When GeoStory find more than one sections/columns it will select the one with at the top (the green one in the screenshot above), see this code:

https://github.com/geosolutions-it/MapStore2/blob/f1b1326e41bde2ff29c7383aefa4c81fc58df76f/web/client/components/geostory/layouts/sections/enhancers/currentPageSectionManager.jsx#L35-L40

In conclusion the navbar highlights the section/column visible in the window view nearest to the top of the story so in case of multiple sections/columns visible in the view the selction could not match the highlight.

made this story to replicate the issue https://dev.mapstore.geo-solutions.it/mapstore/#/geostory/27593/

We should investigate a bit to understand how to solve this behaviour and maybe open a different issue.

**Comment by tdipisa** [comment]
Thank you @allyoucanmap, if I understood correctly your last comment, I think tha we will be able to at least mitigate the issue n° 4 by solving issue n° 5. Is that correct?

**Comment by allyoucanmap** [comment]
@tdipisa Probably it will solve the specific steps described in issue n 5 but the issue it will still be there.
See this other example where loading of contents is not involved:

https://dev.mapstore.geo-solutions.it/mapstore/#/geostory/27593/

![highlight](https://user-images.githubusercontent.com/19175505/90003748-d086ff00-dc94-11ea-85f7-ed5d7139e446.gif)

**Comment by tdipisa** [comment]
Thank you @allyoucanmap, I think that we can leave it as it is for now. We will open a new issue later if needed, let's wait for the PR review first #5767.

**Comment by AGMETEOR** [comment]
For no. 5
------
As there's no way of finding out what size the media will be before hand, we can use an invisible placeholder of a good enough height (500px?) during loading in 2 places
- loading state of `client/components/geostory/common/VisibilityContainer.js`
- loading state of `client/components/media/Image.jsx` / `client/components/media/Video.jsx`

This fixes no. 5
